posts/mathlib-changelog.md

```diff
@@ -0,0 +1,55 @@
+---
+author: "David Chanin"
+category: "Community projects"
+date: 2022-07-28 09:35:23 UTC+02:00
+description: ""
+has_math: false
+link: ""
+slug: mathlib-changelog
+tags: ""
+title: Introducing Mathlib Changelog
+type: text
+---
+
+![mathlib-changelog sample page](/images/changelog_lemma.png)
+
+Tldr; check out [mathlib-changelog.org](https://mathlib-changelog.org) to explore the historical changes to mathlib, and find out what happened to that lemma you were using.
+
+<!-- TEASER_END -->
+
+# A changelog for mathlib
+
+As a beginner in the world of Lean and mathlib (and formal math in general), once I started branching out from the [offical tutorials](https://github.com/leanprover-community/tutorials) and the excellent [Natural Number Game](https://www.ma.imperial.ac.uk/~buzzard/xena/natural_number_game/) and started looking at proofs online, I found that I kept getting confusing errors, even when typing verbatim what I would see on the screen. Searching on Google didn't help, and I often couldn’t find any reference to some of the definitions and lemmas I was seeing in the mathlib docs.
+
+Disoriented, I eventually stumbled onto the amazing [Leanprover Community Zulip](https://leanprover.zulipchat.com/) chat where I learned that mathlib changes very fast, so it’s common for lemmas and theorems to be added and removed all the time. This means mathlib can improve at lightning speed, but also means that content created even a few months ago might not work with new versions of mathlib. After discussing on Zulip the pros and cons of ideas for versioning or maintaining an official changelog, it seemed like the most practical path forward would be to build a tool to automatically extract a searchable changelog from the mathlib git history. Or it could be because I work as a software engineer, so somehow my solution to every problem is to build more software...
+
+The result is [mathlib-changelog.org](https://mathlib-changelog.org), a tool which keeps track of changes to mathlib, and lets you search for any `lemma`, `theorem`, `def`, `abbreviation`, `structure`, or `inductive` which was part of mathlib at any point in the past and track which commits may have modified it. The hope is that this will make it easier to figure out what happened when a theorem you’re using disappears, and what the replacement is.
```

Comment:
I know that unnamed `instance`s are tricky, but why is there no `class`?

Reply:
No real reason, I just wasn't sure what the syntax rules were around `class`. Is it just a simple case of looking for the keyword `class` followed by a name? If so I'll add it to the Changelog

Reply:
Note that for the purpose of the docs, `lemma` and `theorem` are identical. `abbreviation` vs `@[reducible] def` are also equivalent in almost every way, and in practice distinguishing abbreviations from defs provides less value that distinguishing simp lemmas from regular lemmas

Reply:
Just to make sure I understand correctly, you're saying that `lemma` should be deleted from the changelog, and everything that's currently labeled as a `lemma` should instead be labeled as a `theorem`. The same should happen for `abbreviation`, where `abbreviation` should be deleted and instead be relabeled as `def`?

Reply:
Yes, exactly!

Reply:
Good call, should be updated now!

Reply:
> I just wasn't sure what the syntax rules were around class. Is it just a simple case of looking for the keyword class followed by a name?

Yes I think they work the same way as theorems and defs. All are named and get a namespace like thms/defs. They are special `structure`s that support instances, so I think they deserve a separate category.